pkg/services/object/get/util.go

```diff
@@ -158,50 +157,48 @@ func (c *clientWrapper) getObject(exec *execCtx, info coreclient.NodeInfo) (*obj
 			if errors.Is(err, apistatus.ErrObjectAccessDenied) {
 				// Current spec allows other storage node to deny access,
 				// fallback to GET here.
-				obj, err := c.get(exec, key)
+				obj, reader, err := c.get(exec, key)
 				if err != nil {
-					return nil, err
+					return nil, nil, err
 				}
 
-				payload := obj.Payload()
+				pLen := obj.PayloadSize()
 				from := rng.GetOffset()
-				ln := rng.GetLength()
-				if ln == 0 {
-					ln = obj.PayloadSize()
+				var to uint64
+				if ln != 0 {
+					to = from + ln
+				} else {
+					to = pLen
 				}
-				to := from + ln
 
-				if pLen := uint64(len(payload)); to < from || pLen < from || pLen < to {
-					return nil, new(apistatus.ObjectOutOfRange)
+				if to < from || pLen < from || pLen < to {
+					return nil, nil, logicerr.Wrap(apistatus.ErrObjectOutOfRange)
 				}
 
-				return payloadOnlyObject(payload[from:to]), nil
-			}
-			return nil, fmt.Errorf("init payload reading: %w", err)
-		}
-
-		if int64(ln) < 0 {
-			// `CopyN` expects `int64`, this check ensures that the result is positive.
-			// On practice this means that we can return incorrect results for objects
-			// with size > 8_388 Petabytes, this will be fixed later with support for streaming.
-			return nil, new(apistatus.ObjectOutOfRange)
-		}
-
-		ln = min(ln, maxInitialBufferSize)
+				if from > 0 {
+					_, err = io.CopyN(io.Discard, reader, int64(from))
+					if err != nil {
+						return nil, nil, fmt.Errorf("discard %d bytes in stream: %w", from, err)
+					}
+				}
 
-		w := bytes.NewBuffer(make([]byte, ln))
-		_, err = io.CopyN(w, rdr, int64(ln))
-		if err != nil {
-			return nil, fmt.Errorf("read payload: %w", err)
+				return object.New(), struct {
+					io.Reader
+					io.Closer
+				}{
+					Reader: io.LimitReader(reader, int64(to-from)),
+					Closer: reader,
+				}, nil
+			}
+			return nil, nil, fmt.Errorf("init payload reading: %w", err)
 		}
-
-		return payloadOnlyObject(w.Bytes()), nil
+		return object.New(), rdr, nil
```

Comment:
since we have separate return for payload now, suggest to return `rdr` only. But make sure callers won't NPE